intel_mp app instantiations make stats symlink counter forest

wingo opened the pull request:
This commit changes the intel_mp driver to make a forest of symlinks to the shared stats counters, rather than a single link to the shared counters directory.  This allows the app to indicate which queue it's on, as well as allowing ptree and top to naturally track the set of counters used by the app so they can manage associated .rrd files.

As part of this change, we remove the instance-local "pci" directory, instead using per-app "pci" directories.  This allows us to more nicely record which app is on which device.

This commit also updates "snabb top" to get PCI interfaces from within apps' SHM frames.

Fixes #1178.

WDYT @takikawa @dpino ?

wingo committed:
intel_mp app instantiations make stats symlink counter forest

This commit changes the intel_mp driver to make a forest of symlinks to
the shared stats counters, rather than a single link to the shared
counters directory.  This allows the app to indicate which queue it's
on, as well as allowing ptree and top to naturally track the set of
counters used by the app so they can manage associated .rrd files.

As part of this change, we remove the instance-local "pci" directory,
instead using per-app "pci" directories.  This allows us to more nicely
record which app is on which device.

This commit also updates "snabb top" to get PCI interfaces from within
apps' SHM frames.

eugeneia commented on src/core/shm.lua:
```diff
@@ -63,8 +63,7 @@ end
    mkdir(lib.dirname(resolve(name)))
-   assert(S.symlink(root.."/"..resolve(target), root.."/"..resolve(name)),
-          "shm alias failed")
+   assert(S.symlink(root.."/"..resolve(target), root.."/"..resolve(name)))
```
Any reason for removing the assertion error message?

wingo replied:
S.symlink returns two values in an error case, but passing two args to `assert` truncates to the first value.  Removing the message allows the original error to propagate through.

dpino commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -452,6 +509,41 @@ function Intel:new (conf)
    return self
 end
 
+function Intel:create_stats_symlinks ()
```
It seems this function is not used, when are the symlinks created?

eugeneia commented:
On another note, I think having an intel_mp specific directory is a step back on our quest for driver consolidation.

How about the directory was called `/var/run/snabb/pci`, and master processes would symlink `/var/run/snabb/pci/<addr> -> /var/run/snabb/<pid>/group/pci/<addr>`? I think that would solve both of my concerns.

takikawa committed:
Use self.shm frame in get_rxstats / get_txstats

takikawa commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -803,6 +895,7 @@ function Intel:stop ()
       self:unset_pool()
    end
    self:unset_tx_rate()
+   self:remove_stats_symlinks()
```
I think some tests are failing due to this function call. In particular because `self.shm_frame` isn't defined (also above in `create_stats_symlinks`). Tests seem to make progress with this call removed though.

takikawa committed:
Adjust test_10g_counters.sh for new location

takikawa committed:
Remove remove_stats_symlinks call

wingo commented:
@eugeneia regarding "snabb ps", I don't know that there is any significant change -- you get the statistics under the app/ directory of the Snabb instance, and also in the global intel-mp directory, but not in the per-Snabb-instance pci/ directory.

wingo committed:
Merge pull request #1180 from takikawa/intel-mp-symlinks-get-rxstats

Potential fix for failing unit test on #1179

wingo committed:
Remove unused symlink creation routines

These functions were part of an earlier patch that didn't use the
automatic management of the shm tree via core.app infra.

wingo committed:
Minor formatting change

wingo committed:
Update l2vpn for intel_mp stats change

wingo committed:
Remove extraneous diff

wingo committed:
Remove more extraneous diff

wingo committed:
Revert "Minor formatting change"

Make CI run

wingo commented:
I have no idea why these tests are failing :(((  Deep flakiness :(((

takikawa commented:
@wingo Re: the latest test failures, my theory after poking around a bit is that one problem is with the `return counter.open(target(name))` line in my commit 79d45461171f4ee34b4d0ed064e1f8941b7cd87c.

The issue with that call is that the target may not actually be created yet, if the snabb apps get stopped and restarted (as happens in the snabbnfv selftest).

For example, say we restart the stat collecting NIC app. When it closes, it will delete the shm frame for the stats and relinquish its stat collection status. When it comes back up, it may see that the lock is still held for the NIC (because of another NIC app) and won't start back up as stat collector.

In that case, the `self.shm` counters can't actually open the symlinked targets (since they were deleted and not recreated) and the calls fail.

Basically, the interaction between stat collection and restarting apps seems messy right now. I think we could change the line to only open the counter if it `shm.exists()` (and otherwise return an object that will error when used or something), or maybe make `get_rxstats` read the counter a different way.

wingo committed:
intel_mp: Don't eagerly open shared counters

Instead, open them when and if they are read.

wingo commented:
Latest commit avoids eagerly opening the counters.  Let's see what happens!

wingo commented:
yes!!! @takikawa wdyt?
